torch/distributed/pipelining/stage.py

```diff
@@ -815,15 +820,8 @@ def backward_one_chunk(
 
         self.bwd_cache[bwd_chunk_id] = grads_input
 
-        if self.is_last and not self.is_first:
-            # Autograd dependencies:
-            #    rest_of_autograd_graph -> stage_output -> loss
-            # stage_output is no longer used in the last stage for backward and only needed
-            # to return to the user in merge_output_chunks, therefore
-            # this should be detached to release autograd graph context and free memory earlier
-            for t in stage_output:
-                if not t._is_view():  # views are not detachable in-place
-                    t.detach_()
+        # stage_output is freed after this function returns and
```

Comment:
I think we still need to detach here, at least for the last stage where `stage_output` is actually the loss (https://github.com/pytorch/pytorch/blob/main/torch/distributed/pipelining/stage.py#L758). We want to detach the loss after backward so that the memory can be freed earlier.

If we perform the detach as above in `schedules.py::_update_losses`, this is only called at the end of the step, so all losses and subsequentially the entire autograd graph will be help until the end of the step.

Reply:
that's true, in that case there should be no issue with detaching the loss value in-place